src/plugins/EnergyRAPL.cpp

```diff
@@ -0,0 +1,154 @@
+	_microJoulesPath = microJoulesPath.str();
+	// Value range for current micro joules
+	std::ostringstream rangeMicroJoulesPath;
+	rangeMicroJoulesPath << domainBasePath << "/max_energy_range_uj";
+	std::ifstream rangeMicroJoulesFile(rangeMicroJoulesPath.str());
+	rangeMicroJoulesFile >> _rangeMicroJoules;
+	reset();
+}
+
+void EnergyRAPL::RAPLCounter::reset() {
+	// Update last micro joules
+	update();
+	_microJoules = 0;
+}
+
+double EnergyRAPL::RAPLCounter::update() {
+	long long currentMicroJoules;
+	std::ifstream packageIdFile(_microJoulesPath);
+	packageIdFile >> currentMicroJoules;
+	long long deltaMicroJoules = currentMicroJoules - _lastMicroJoules;
+	// Correct counter overflow (occurs around every 60 seconds)
+	if (0 > deltaMicroJoules) {
+		deltaMicroJoules += _rangeMicroJoules;
+	}
+	_lastMicroJoules = currentMicroJoules;
+	_microJoules += deltaMicroJoules;
+	return ((double)_microJoules) / 1000 / 1000;
+}
+
+int EnergyRAPL::getNumberOfPackages() {
+	int maxPackageIdx = 0;
+	for (int cpuIdx = 0;; cpuIdx++) {
+		std::ostringstream packageIdxPath;
+		packageIdxPath << "/sys/devices/system/cpu/cpu" << cpuIdx << "/topology/physical_package_id";
+		std::ifstream packageIdFile(packageIdxPath.str());
+		if (!packageIdFile.good()) break;
+		int packageIdx = -1;  // No more packages
+		packageIdFile >> packageIdx;
+		if (packageIdx > maxPackageIdx) maxPackageIdx = packageIdx;
+	}
+	return maxPackageIdx + 1;
+}
+
+void EnergyRAPL::init(ParticleContainer* particleContainer, DomainDecompBase* domainDecomp, Domain* domain) {
+#ifdef ENABLE_MPI
+	MPI_Get_processor_name(_processorName, &_processorNameLength);
+	MPI_Comm_rank(MPI_COMM_WORLD, &_thisRank);
+#endif
+	std::ostringstream outputFilename;
+	outputFilename << _outputprefix << ".tsv";
+	std::ofstream outputFile(outputFilename.str().c_str());
+	outputFile << "milliseconds\tsimstep\tjoules" << std::endl;
+	// For each package...
+	int maxPackageIdx = 0;
+	int numberOfPackages = getNumberOfPackages();
+	for (int packageIdx = 0; packageIdx < numberOfPackages; packageIdx++) {
+		std::ostringstream packageBasePath;
+		packageBasePath << _basePathRAPL << "intel-rapl:" << packageIdx;
+		// add package domain and...
+		Log::global_log->info() << "[" << getPluginName() << "] Adding package domain " << packageBasePath.str()
+								<< std::endl;
+		_counters.push_back(RAPLCounter(packageBasePath.str()));
+		// for each domain in package...
+		for (int domainIdx = 0;; domainIdx++) {
+			std::ostringstream domainBasePath;
+			domainBasePath << packageBasePath.str() << "/intel-rapl:" << packageIdx << ":" << domainIdx;
+			std::ostringstream domainNamePath;
+			domainNamePath << domainBasePath.str() << "/name";
+			std::string domainName;
+			std::ifstream domainNameFile(domainNamePath.str());
+			if (!domainNameFile.good()) break;  // (if domain exists...)
+			domainNameFile >> domainName;
+			if (0 != domainName.compare("dram")) continue;  // (and is DRAM domain)
```

Comment:
What is the rational behind this choice?
Is this safe/portable? The Intel documentation states only for platforms targeting the server segment that the DRAM domain is supported.

Reply:
Yes, this is documented in section 15.10.2 of the Intel Architectures Software Developer’s Manuals Volume 3, but energy consumption should be most relevant on server architecture (cluster compute nodes) anyway.
On client (and AMD) CPUs this results in only measuring the energy consumption of the package domain.
The plugin also reports the used RAPL domains to the log (s. L78,93).
Energy measurements from RAPL cannot be used _in a vacuum_, and only the package and DRAM domains should be relevant for ls1.
I will make this explicit in the plugin documentation.